datafusion/expr/src/expr.rs

```diff
@@ -1674,6 +1674,69 @@ impl Expr {
     }
 }
 
+impl Normalizeable for Expr {
+    fn can_normalize(&self) -> bool {
+        #[allow(clippy::match_like_matches_macro)]
+        match self {
+            Expr::BinaryExpr(BinaryExpr {
+                op:
+                    _op @ (Operator::Plus
+                    | Operator::Multiply
+                    | Operator::BitwiseAnd
+                    | Operator::BitwiseOr
+                    | Operator::BitwiseXor
+                    | Operator::Eq
+                    | Operator::NotEq),
+                ..
+            }) => true,
+            _ => false,
+        }
+    }
+}
+
+impl NormalizeEq for Expr {
+    fn normalize_eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (
+                Expr::BinaryExpr(BinaryExpr {
+                    left: self_left,
+                    op: self_op,
+                    right: self_right,
+                }),
+                Expr::BinaryExpr(BinaryExpr {
+                    left: other_left,
+                    op: other_op,
+                    right: other_right,
+                }),
+            ) => {
+                if self_op != other_op {
+                    return false;
+                }
+
+                if matches!(
+                    self_op,
+                    Operator::Plus
+                        | Operator::Multiply
+                        | Operator::BitwiseAnd
+                        | Operator::BitwiseOr
+                        | Operator::BitwiseXor
+                        | Operator::Eq
+                        | Operator::NotEq
+                ) {
+                    (self_left.normalize_eq(other_left)
+                        && self_right.normalize_eq(other_right))
+                        || (self_left.normalize_eq(other_right)
+                            && self_right.normalize_eq(other_left))
+                } else {
+                    self_left.normalize_eq(other_left)
+                        && self_right.normalize_eq(other_right)
+                }
+            }
+            (_, _) => self == other,
```

Comment:
Hmm, this actually means that if there is an other expr above the cumulative `BinaryExpr` operation, then we don't normalize? E.g. `Not(a == b)` is not semantically equals to `Not(b == a)`?

Reply:
yeah, I got it, I will add more arm for this scope later.

Reply:
Hi @peter-toth, Apologies for the delayed commit. I've added more arm in the `normalize_eq` function to handle cumulative `BinaryExpr` comparisons for other expressions. While working on this, I also noticed that other expressions could benefit from normalization. For example, with the `InList` and `CaseWhen` expression, we can ignore the order of elements.

You can see the relevant code here:
https://github.com/apache/datafusion/blob/cc11692226da7e5dd49caaee2a8c3e66af920d4c/datafusion/expr/src/expr.rs#L2013
https://github.com/apache/datafusion/blob/cc11692226da7e5dd49caaee2a8c3e66af920d4c/datafusion/expr/src/expr.rs#L2034-L2036

In this case, I think the normalize_eq(&self, other: &Self) -> bool trait is not the best way to handle this scenario for almost exponential time complexity. At this moment, it's a good idea to normalize it first and then compare it. Do you have any suggestions on how to approach this?

Reply:
Yeah, that is not easy to handle now, but IMO this PR looks great without that as well.

Actually, I'm working on something that stores and updates certain statistics and properties (like the hash) of nodes automatically during transformations. I think once the hashes of a node's children will be cheap we can use it to sort the children and call `normalize_eq()` on the pairs.

Reply:
> Yeah, that is not easy to handle now, but IMO this PR looks great without that as well.
> 
> Actually, I'm working on something that stores and updates certain statistics and properties (like the hash) of nodes automatically during transformations. I think once the hashes of a node's children will be cheap we can use it to sort the children and call `normalize_eq()` on the pairs.

Yeah, I agree store and update hash for nodes greatly. Now, I am handle hash computing cumulativly and `normalize_eq` separately.

1. computing node hash here:
https://github.com/apache/datafusion/blob/cc11692226da7e5dd49caaee2a8c3e66af920d4c/datafusion/common/src/cse.rs#L325-L331

2. invoke `normalize_eq` here:
https://github.com/apache/datafusion/blob/cc11692226da7e5dd49caaee2a8c3e66af920d4c/datafusion/common/src/cse.rs#L396-L399

Reply:
Yes, that's the best we can do now. But I will try to open a PR in a few days.